### ISSUE TITLE
L025: Not recognizing/handling unnests properly

### ISSUE THREAD

**Issue by davajm** [sqlfluff/sqlfluff#3946]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

We had a model break after applying `sqlfluff fix` where was complaining that an alias never was used, but it was. I managed to simplify the issue into two small examples.

Assume that we have these two models:
`a.sql`
```
with source_model as (
    select [1, 2, 3] as nested_field
    union all
    select [1, 2] as nested_field
)

select nf from source_model as sm, unnest(sm.nested_field) as nf

```

`b.sql`
```
with source_model as (
    select [1, 2, 3] as nested_field
    union all
    select [1, 2] as nested_field
)

select (select count(*) from unnest(sm.nested_field)) as cnt
from source_model as sm

```

### Expected Behaviour

For `a.sql` I guess `sqlfluff fix` should either 
1. Remove the `sm` alias from both the table **and** the unnest, or
2. Not report any violations at all

For `b.sql` I believe no violations should be reported as the alias is used inside the subquery.

### Observed Behaviour

`a.sql`
```
== [models/a.sql] FAIL
L:   7 | P:  32 | L025 | Alias 'sm' is never used in SELECT statement.
==== fixing violations ====
1 fixable linting violations found
Are you sure you wish to attempt to fix these? [Y/n]
```
Here it (correctly?) identifies the violation, but the fix ends up producing a broken model:
```
with source_model as (
    select [1, 2, 3] as nested_field
    union all
    select [1, 2] as nested_field
)

select nf from source_model, unnest(sm.nested_field) as nf

```

`b.sql`
```
== [models/b.sql] FAIL
L:   8 | P:  22 | L025 | Alias 'sm' is never used in SELECT statement.
==== fixing violations ====
1 fixable linting violations found
Are you sure you wish to attempt to fix these? [Y/n]
```
Here it **incorrectly** identifies the violation and produces a broken model:
```
with source_model as (
    select [1, 2, 3] as nested_field
    union all
    select [1, 2] as nested_field
)

select (select count(*) from unnest(sm.nested_field)) as cnt
from source_model

```

### How to reproduce

Run `sqlfluff fix` on the two models above and see the result.

### Dialect

BigQuery

### Version

sqlfluff, version 1.3.2
sqlfluff-templater-dbt, version: 1.3.2
Python, version 3.9.13

### Configuration

`.sqlfluff`
```
[sqlfluff]
dialect = bigquery
templater = dbt

[sqlfluff:templater:jinja]
apply_dbt_builtins = True
```

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by keraion** [comment]
Fixed by #5726
